okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java

```diff
@@ -170,32 +172,44 @@ private RealConnection findConnection(int connectTimeout, int readTimeout, int w
         refusedStreamCount = 0;
       }
     }
-    RealConnection newConnection = new RealConnection(connectionPool, selectedRoute);
+
+    RealConnection result = new RealConnection(connectionPool, selectedRoute);
 
     synchronized (connectionPool) {
-      acquire(newConnection);
-      Internal.instance.put(connectionPool, newConnection);
-      this.connection = newConnection;
+      acquire(result);
+      Internal.instance.put(connectionPool, result);
```

Comment:
I’ve added code to defer pooling. I think it’s a net win.

I haven’t yet done the simplifying step. I’ll do that in follow-up.